compiler/optimizer/UnsafeSubexpressionRemover.cpp

```diff
@@ -0,0 +1,196 @@
+   //
+   // 1. Stores can have multiple children -- as many as three for
+   //    an iwrtbar.  If the store is turning into a nop, we must
+   //    anchor all the other children.
+   // 2. Stores can appear under other nodes like CHKs and
+   //    compressedRefs.  Particularly for the CHKs, we must
+   //    perform the appropriate checks despite eliminating the
+   //    store, and so we must leave the store node in the
+   //    appropriate shape so its parents will do the right thing.
+   // 3. Some expressions are unsafe to evaluate after stores are
+   //    eliminated, and those expressions themselves must NOT be
+   //    anchored.
+   //
+   // In these transformations, we adopt the convention that we
+   // retain the original store whenever possible, modifying it
+   // in-place, and we retain its first child if possible.  These
+   // conventions are sometimes required, and so we just adopt them
+   // universally for the sake of consistency, efficiency, and
+   // readability in the jit logs.
+   TR::Node *node = treeTop->getNode();
+
+   // Anchor all children but the first
+   //
+   for (int32_t i = 1; i < node->getNumChildren(); i++)
+      {
+      TR::Node *child = node->getChild(i);
+      anchorIfSafe(child, treeTop);
+      child->recursivelyDecReferenceCount();
+      }
+
+   node->setNumChildren(1);
+   //
+   // Note that the node at this point could be half-baked: an indirect
+   // store or write barrier with one child.  This is an intermediate
+   // state, and further surgery is required for correctness.
+
+   // Eliminate the store
+   //
+   if (treeTop->getNode()->getOpCode().isSpineCheck() && treeTop->getNode()->getFirstChild() == node)
+      {
+      // SpineCHKs are special. We don't want to require all the
+      // opts and codegens to do the right thing with a PassThrough
+      // child.  Hence, we'll turn the tree into a const, which is
+      // a state the opts and codegens must cope with anyway,
+      // because if this tree had been a load instead of a store,
+      // the optimizer could have legitimately turned it into a const.
+      //
+      TR::Node *child = node->getChild(0);
+      anchorIfSafe(child, treeTop);
+      child->recursivelyDecReferenceCount();
+      TR::Node::recreate(node, comp()->il.opCodeForConst(node->getSymbolReference()->getSymbol()->getDataType()));
+      node->setFlags(0);
+      node->setNumChildren(0);
+      }
+   else
+      {
+      // Vandalize the first child if it's unsafe to evaluate
+      //
+      TR::Node *child = node->getFirstChild();
+      anchorSafeChildrenOfUnsafeNodes(child, treeTop);
+      if (isUnsafe(child))
+         {
+         // Must eradicate the dead expression.  Can't leave it
+         // anchored because it's not safe to evaluate.
+         //
+         child->recursivelyDecReferenceCount();
+         TR::Node *dummyChild = node->setAndIncChild(0, TR::Node::createConstDead(child, TR::Int32, 0xbad1 /* eyecatcher */));
```

Comment:
That's a good question.  The code below recreates the top-level node as either a `PassThrough` or a `treetop` node.  If the top-level node's reference count was zero, it becomes a `treetop`, with the vandalized constant as its only child, so that's safe.

However, if the top-level node's reference count was greater than zero, the store might be, as noted by point 2 in the comment above, a commoned child of a `CHK` node perhaps or a `compressedrefs`.  I'm not sure what will happen during later optimization passes or code generation with something like this.

```
  NULLCHK  or  compressedrefs
    PassThrough
      iconst  0xbad1dead
```

I can rework this to ensure the type of the constant matches whatever type that child node previously represented.